dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/persistence/datastore/FileDataStoreHandler.kt

```diff
@@ -0,0 +1,289 @@
+            dataStoreDirectory = dataStoreDirectory,
+            dataStoreFileName = dataStoreFileName
+        )
+
+        if (!datastoreFile.existsSafe(internalLogger)) {
+            return null // no datastore file found
+        }
+
+        return readFromDataStoreFile(datastoreFile, deserializer, fileTLVBlockReader, version)
+    }
+
+    @WorkerThread
+    private fun writeToFile(dataStoreFile: File, data: ByteArray) {
+        fileReaderWriter.writeData(
+            file = dataStoreFile,
+            data = data,
+            append = false
+        )
+    }
+
+    @WorkerThread
+    private fun <T : Any> getDataBlock(
+        data: T,
+        serializer: Serializer<T>
+    ): ByteArray? {
+        val serializedData = serializer.serialize(data)?.toByteArray()
+
+        if (serializedData == null) {
+            logFailedToSerializeDataError()
+            return null
+        }
+
+        val dataBlock = TLVBlock(
+            type = TLVBlockType.DATA,
+            data = serializedData
+        )
+
+        return dataBlock.serialize()
+    }
+
+    @WorkerThread
+    private fun getLastUpdateDateBlock(): ByteArray? {
+        val now = System.currentTimeMillis()
+        val lastUpdateDateByteArray = now.toByteArray()
+        val lastUpdateDateBlock = TLVBlock(
+            type = TLVBlockType.LAST_UPDATE_DATE,
+            data = lastUpdateDateByteArray
+        )
+
+        return lastUpdateDateBlock.serialize()
+    }
+
+    @WorkerThread
+    private fun getVersionCodeBlock(): ByteArray? {
+        val versionCodeByteArray = CURRENT_DATASTORE_VERSION.toByteArray()
+        val versionBlock = TLVBlock(
+            type = TLVBlockType.VERSION_CODE,
+            data = versionCodeByteArray
+        )
+
+        return versionBlock.serialize()
+    }
+
+    @WorkerThread
+    @Suppress("ReturnCount")
+    private fun <T : Any> readFromDataStoreFile(
+        datastoreFile: File,
+        deserializer: Deserializer<String, T>,
+        fileTLVBlockReader: FileTLVBlockReader,
+        requestedVersion: Int
+    ): T? {
+        val tlvBlocks = fileTLVBlockReader.all(datastoreFile)
+
+        // there should be as many blocks read as there are block types
+        if (tlvBlocks.size != TLVBlockType.values().size) {
+            logInvalidNumberOfBlocksError(tlvBlocks.size)
+            return null
+        }
+
+        val dataStoreContents = tryToMapToDataStoreContents(deserializer, tlvBlocks)
+            ?: return null
+
+        val fileVersionIsWrong = dataStoreContents.versionCode != requestedVersion
+        val fileIsTooOld = isDataStoreTooOld(dataStoreContents.lastUpdateDate)
+
+        // the version check is a double redundancy. Since we store each file in a folder
+        // whose name contains the version it should be impossible to read a file
+        // that contains the wrong version.
+        if (fileVersionIsWrong) {
+            logInvalidVersionError()
+        }
```

Comment:
why do we need it then?

Reply:
I'm moving the version validation logic to the calling feature. To align with ios, version will be different for the directory and for the entries - if read is called without specifying the version then data will always be returned if an entry with that key is found. If version is specified then if the entry has a different version no data will be returned. When writing, specifying a version will write with that version, not specifying a version will write the entry with version 0 (even if it downgrades from a previous higher version number on the entry)